Bluetooth: host: Add API to HCI CTE conn request enable procedure

ppryga-nordic opened the pull request:
There were no implementation for HCI_LE_Connection_CTE_Request_Enable
command from BT 5.3 Core specification.

The PR adds implementation and API to be used by applications.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: host: Add API to HCI CTE conn request enable procedure

There were no implementation for HCI_LE_Connection_CTE_Request_Enable
command from BT 5.3 Core specification.

The PR adds implementation and API to be used by applications.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -142,6 +142,19 @@ struct bt_df_conn_iq_samples_report {
 };
+
+struct bt_df_conn_cte_req_params {
+	/** Requested interval for initiating the CTE Request procedure. Value 0x0 means, run the
+	 * procedure once. Other values are intervals in number of connection events, to run the
+	 * command periodically.
+	 */
```
That's quite a long brief. Consider
```suggestion
	/** 
	* @brief Requested interval for initiating the CTE Request procedure. 
	* 
	* Value 0x0 means, run the
	 * procedure once. Other values are intervals in number of connection events, to run the
	 * command periodically.
	 */
```

Is `interval` the name used by the spec? Otherwise I think it is misguiding since interval is usually expressed in a time unit, but this is number of connection events.

ppryga replied:
`interval` refers to `CTE_Request_Interval` from the BT 5.3 Core spec.

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -142,6 +142,19 @@ struct bt_df_conn_iq_samples_report {
+	uint8_t interval;
+	/* Requested length of the CTE */
+	uint8_t cte_length;
+	/* Requested type of the CTE */
+	uint8_t cte_type;
```
Reference to possible values?

ppryga replied:
Added list of allowed values.

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -217,4 +230,23 @@ int bt_df_conn_cte_rx_enable(struct bt_conn *conn, const struct bt_df_conn_cte_r
  */
 int bt_df_conn_cte_rx_disable(struct bt_conn *conn);
 
+/**
+ * @brief Enable Constant Tone Extension request procedure for a connection.
+ *
+ * @param conn   Connection object.
+ * @param params CTE receive and sampling parameters.
+ *
+ * @return Zero in case of success, other value in case of failure.
+ */
+int bt_df_conn_cte_req_enable(struct bt_conn *conn, const struct bt_df_conn_cte_req_params *params);
+
+/**
+ * @brief Disable Constant Tone Extension request procedure for a connection.
+ *
+ * @param conn   Connection object.
+ *
+ * @return Zero in case of success, other value in case of failure.
+ */
+int bt_df_conn_cte_req_disable(struct bt_conn *conn);
+
```
This isn't something we've done a lot in the API, but if functions are guarded by specific KConfigs (BT_DF_CONNECTION_CTE_REQ), we could add a reference to the Kconfig (`@kconfig{CONFIG_BT_DF_CONNECTION_CTE_REQ`)

ppryga replied:
Added info: `* The function is available if @kconfig{CONFIG_BT_DF_CONNECTION_CTE_REQ is enabled.`

ppryga replied:
Documentation build complains about unknown command @kconfig, I've changed the lines to:
`* The function is available if Kconfig CONFIG_BT_DF_CONNECTION_CTE_REQ is enabled.`

Thalley replied:
> 

Hmm, that's odd. We use it e.g. for `bt_enable`: 

```
/**
 * @brief Enable Bluetooth
 *
 * Enable Bluetooth. Must be the called before any calls that
 * require communication with the local Bluetooth hardware.
 *
 * When @kconfig{CONFIG_BT_SETTINGS} has been enabled and the application is not
 * managing identities of the stack itself then the application must call
 * @ref settings_load() before the stack is fully enabled.
 * See @ref bt_id_create() for more information.
 *
 * @param cb Callback to notify completion or NULL to perform the
 * enabling synchronously.
 *
 * @return Zero on success or (negative) error code otherwise.
 */
int bt_enable(bt_ready_cb_t cb);
```

So if it fails, it looks like something else is wrong.

ppryga replied:
I've added @kconfig again, lets see the output from doc-build.

Thalley replied:
Seems to work :)

Thalley commented on subsys/bluetooth/host/direction.c:
```diff
@@ -593,6 +598,82 @@ int hci_df_prepare_connection_iq_report(struct net_buf *buf,
+	return true;
+}
+
+static void prepare_conn_cte_req_enable_cmd_params(struct net_buf *buf, struct bt_conn *conn,
```
```suggestion
static void prepare_conn_cte_req_enable_cmd_params(struct net_buf *buf, const struct bt_conn *conn,
```

ppryga replied:
`const` added.

Thalley commented on subsys/bluetooth/host/direction.c:
```diff
@@ -593,6 +598,82 @@ int hci_df_prepare_connection_iq_report(struct net_buf *buf,
+	struct net_buf *buf, *rsp;
+	int err;
+
+	if (enable) {
+		if (!validate_cte_req_params(conn, params->cte_length, params->cte_type)) {
+			return -EINVAL;
+		}
+	}
```
```suggestion
	if (enable && !validate_cte_req_params(conn, params->cte_length, params->cte_type)) {
		return -EINVAL;
	}
```

ppryga replied:
Change added.

Thalley commented on subsys/bluetooth/host/direction.c:
```diff
@@ -767,3 +848,50 @@ int bt_df_conn_cte_rx_disable(struct bt_conn *conn)
+	CHECKIF(!conn)
+	{
+		return -EINVAL;
+	}
+	CHECKIF(!params)
```
```suggestion
	}
	
	CHECKIF(!params)
```

ppryga replied:
Added empty line

Thalley reviewed (changes requested):
Minor comments

ppryga commented:
Comments addressed! Thank you @Thalley for lightening fast review!

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -142,6 +142,25 @@ struct bt_df_conn_iq_samples_report {
+	/**
+	 * @brief Requested type of the CTE.
+	 *
+	 * Allowed values are: BT_HCI_LE_AOA_CTE, BT_HCI_LE_AOD_CTE_1US, BT_HCI_LE_AOD_CTE_2US.
+	 */
+	uint8_t cte_type;
```
Something to consider, but not completely necessary: We've often `define`d values in the header files rather than relying on the values defined in `hci.h`, see e.g. `gap.h` which mostly consists of such files. 

It could be beneficial to define values such as `BT_DF_AOA_CTE` and `BT_DF_CTE_1US` in this file, so that the user does not need to include `hci.h`.

ppryga replied:
I've added separate commit because there was an `enum by_df_cte_type`. The former one was changed to `enum by_df_cte_type_bits` and all members has added suffix `_BIT` or `_BITS` (for multi-bit value).

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -9,20 +9,35 @@
 enum bt_df_cte_type {
+	/** Angle of Arrival mode. Antenna switching done on receiver site. */
+	BT_DF_CTE_TYPE_AOA = 0,
+	/** Angle of Departure mode with 1 us antenna switching slots.
+	 *  Antenna switching done on transmitter site.
+	 */
+	BT_DF_CTE_TYPE_AOD_1US = 1,
```
Consider the following to avoid having long `brief`s:

```suggestion
	/** 
	 * @brief Angle of Departure mode with 1 us antenna switching slots.
	 *
	 * Antenna switching done on transmitter site.
	 */
	BT_DF_CTE_TYPE_AOD_1US = 1,
```

(same comment applies below as well)

ppryga replied:
Corrected

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -9,20 +9,35 @@
+};
+
+/** Constant Tone Extension (CTE) types for use as bit field values */
+enum bt_df_cte_type_bits {
```
Hmm, not sure I am a fan of having both `bt_df_cte_type` and `bt_df_cte_type_bits`, if they define the same values as a bitfield and just as values. 

Do you think it would be possible to keep only of them, and modify the API to somehow only use one of them? Similar to how we use the `BT_GAP_LE_PHY` from `gap.h`.

ppryga replied:
Well, I should have stated my concerns before posting changes. In general I agree. 
I think that public API can use single `enum` just differentiate the usage. E.g. for `cte_type` use it as single value member; whereas for `cte_types` use it as bitfiled.

Are we have a common opinion about it?

Thalley replied:
> Do we have a common opinion about it?

I think so, not sure I follow your comment 100% to be honest :) But it sounds like what you are proposing is the same solution we have for the `BT_GAP_LE_PHY` values

ppryga replied:
Changes done. I've aligned other code to use the same approach.

ppryga-nordic committed:
Bluetooth: host: Align usage of cte_type in DF host implementation

The CTE type is used in two ways by HCI layer:
1) single value representing particular CTE type: AoA, AoD 1 us,
  AoD 2 us
2) bit-filed where bits 0-2 represent particular CTE types AoA
  AoD 1 us, AoD 2 us

The bit-field is used to inform Controller about allowed types
of CTE, hence single value carries more than one value.
To avoid confusion between these use cases in code that refers
to case 1) all named cte_type (singular form). For case 2)
cte_types (plural form) is used.

There is an enumeration that is used for both cases:
bt_df_cte_type. For cte_type only single value from the
enumeration may be assigned to variable except
BT_DF_CTE_TYPE_NONE and BT_DF_CTE_TYPE_ALL.
For cte_types all enum members may be used. Ocasionally
BT_DF_CTE_TYPE_NONE may be excluded. If that is true,
it is described in code documentation.

Thanks to that applications are released from requirement
to include hci.h header file.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
@Thalley looks that all comments were addressed.

ppryga commented:
@jhedberg @alwa-nordic could you take a look?

Thalley commented on include/bluetooth/direction.h:
```diff
@@ -76,8 +85,12 @@ struct bt_df_adv_cte_tx_param {
+	/**
+	 * @brief Bitfield with allowed CTE types.
+	 *
+	 *  Allowed values are defined by @ref bt_df_cte_type, except BT_DF_CTE_TYPE_NONE.
```
Is there a usecase for `BT_DF_CTE_TYPE_NONE`?

ppryga replied:
It was added purposely in the past, but it looks that it is not necessary anymore. 
I'll create a separate PR to remove it.

Thalley reviewed (approved):
Minor last question, otherwise LGTM

ppryga commented:
Rebased on main due to conflicts.

ppryga commented:
@carlescufi could you take a look on the PR?

carlescufi reviewed (approved):
LGTM

jhedberg commented on subsys/bluetooth/host/direction.c:
```diff
@@ -593,6 +610,78 @@ int hci_df_prepare_connection_iq_report(struct net_buf *buf,
 #endif /* CONFIG_BT_DF_CONNECTION_CTE_RX */
 
+#if defined(CONFIG_BT_DF_CONNECTION_CTE_REQ)
+static bool validate_cte_req_params(const struct bt_conn *conn, uint8_t cte_type,
```
I'm not sure the name of this function is quite ideal. Since it returns bool the name should IMO be a boolean assertion (now the form of the verb makes it a request). I'd just follow similar naming as we've used elsewhere for parameter validation.

Thalley replied:
Agreed. 

Suggest to follow the naming convention used by e.g. 
```
static bool valid_adv_ext_param(const struct bt_le_adv_param *param)
```
thus renaming this to 
```
static bool valid_cte_req_params(const struct bt_le_adv_param *param)
```

jhedberg replied:
Yeah, that's fine with me

ppryga replied:
Changed according to @Thalley proposal.
I've also aligned other functions names `validate_**` in additional commit.

ppryga-nordic committed:
Bluetooth: host: Style correction. Change names validate_** to valid_**

Align to naming convention for function responsible for HCI commands
parameters validation.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
@jhedberg comment addressed.
Also rebased to resolve conflicts.
